sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DefaultSynchronousMetricStorage.java

```diff
@@ -108,17 +105,38 @@ public void recordDouble(double value, Attributes attributes, Context context) {
               + ". Dropping measurement.");
       return;
     }
-    Lock readLock = aggregatorHolder.lock.readLock();
-    readLock.lock();
+    AggregatorHolder<T, U> aggregatorHolder = getHolderForRecord();
     try {
       AggregatorHandle<T, U> handle =
           getAggregatorHandle(aggregatorHolder.aggregatorHandles, attributes, context);
       handle.recordDouble(value, attributes, context);
     } finally {
-      readLock.unlock();
+      aggregatorHolder.recordsInProgress.addAndGet(-2);
     }
   }
 
+  /**
+   * Obtain the AggregatorHolder for recording measurements, re-reading the volatile
+   * this.aggregatorHolder until we access one where recordsInProgress is even. Collect sets
+   * recordsInProgress to odd as a signal that AggregatorHolder is stale and is being replaced.
+   * Record operations increment recordInProgress by 2. Callers MUST decrement recordsInProgress by
+   * -2 when complete as a signal to Collect that record operations are active and must complete
+   * before its safe to collect.
+   */
+  private AggregatorHolder<T, U> getHolderForRecord() {
+    do {
+      AggregatorHolder<T, U> aggregatorHolder = this.aggregatorHolder;
+      int recordsInProgress = aggregatorHolder.recordsInProgress.addAndGet(2);
+      if (recordsInProgress % 2 == 0) {
+        return aggregatorHolder;
+      } else {
+        // Collect is in progress, decrement recordsInProgress to allow collect to proceed and
+        // re-read aggregatorHolder
+        aggregatorHolder.recordsInProgress.addAndGet(-2);
+      }
+    } while (true);
```

Comment:
so the key is that this loop will at most loop twice, because if `recordsInProgress` is odd then the collect thread must have already updated `this.aggregatorHolder` to the next one?

Reply:
Correct.